pkg/tsdb/prometheus/prometheus.go

```diff
@@ -283,13 +291,20 @@ func (s *Service) parseQuery(queryContext *backend.QueryDataRequest, dsInfo *Dat
 		expr = strings.ReplaceAll(expr, "$__range", strconv.FormatInt(rangeS, 10)+"s")
 		expr = strings.ReplaceAll(expr, "$__rate_interval", intervalv2.FormatDuration(calculateRateInterval(interval, dsInfo.TimeInterval, s.intervalCalculator)))
 
+		queryType := Range
+
+		// Align query range to step. It rounds start and end down to a multiple of step.
+		start := int64(math.Floor((float64(query.TimeRange.From.Unix()+model.OffsetSec)/interval.Seconds()))*interval.Seconds() - float64(model.OffsetSec))
+		end := int64(math.Floor((float64(query.TimeRange.To.Unix()+model.OffsetSec)/interval.Seconds()))*interval.Seconds() - float64(model.OffsetSec))
```

Comment:
```suggestion
start := int64(math.Floor(float64(query.TimeRange.From.Unix()+model.OffsetSec - model.OffsetSec)))
end := int64(math.Floor(float64(query.TimeRange.To.Unix()+model.OffsetSec - model.OffsetSec)))
```

Reply:
I may have to try and fetch my maths knowledge from the upper shelf, but couldn't this be simplified since we are dividing and multiplying by the same factor? 😬

Reply:
This is basically useful when `(from + offset) / interval` ends up being decimal number and we use math.floor on it.  So when I remove type conversions and use specific example:

```
Math.Floor((1631784735 + 120) / 30) * 30 - 120
Math.Floor(54392828.5) * 30 - 120
54392828 * 30 - 120
1631784840 - 120
1631784720
```

The same logic is currently used on frontend https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/datasource.ts#L843-L844

Reply:
Hm, okay, so by doing that we have a _15sec_ difference compared to if we just simplified the equation (`1631784720` vs `1631784735`). Is this desired?

Reply:
Your simplified version is basically `math.Floor(query.TimeRange.From.Unix() + model.OffsetSec - model.OffsetSec)` where `model.OffsetSec - model.OffsetSec = 0 ` so it would be just `math.Floor(query.TimeRange.From.Unix())`. 

But yeah, 15sec difference is desired because it aligns timerange to step. If you run `rate(ALERTS[1m])` and select Last 3h time range and update step to 1h it is desired that each data points starts at the beginning of each hour so x:00:00. If we didn't aligned it this way, it could start at any time - e.g. x:25:32.

Reply:
Here is more context https://github.com/grafana/grafana/pull/16110

Reply:
TIL then, thanks for the detailed explanation @ivanahuckova!